### ISSUE TITLE
Discussion on design of domain and grid interface

### ISSUE THREAD

**Issue by jlperla** [JuliaDiffEq/PDERoadmap#4]
This issue isolates and continues discussions in https://github.com/JuliaDiffEq/DifferentialEquations.jl/issues/260 with @dlfivefifty and @ChrisRackauckas  on a joint design of the interface for domains, etc. between the ApproxFun, DiffEqOperators, and ModelingToolkit libraries.  We hope to implement it in the prototype: https://github.com/JuliaDiffEq/PDERoadmap/pull/1

Also see https://github.com/JuliaDiffEq/ModelingToolkit.jl/blob/master/test/domains.jl

For a starting point in one dimension,
```
d = Interval(1.0, 10.0) #Continuous domain
x = RegularGrid(d, N)
```
Note: I think it is better to give points than dx, but I am open.  The reason is that if you give a `dx` then you never know how well things line up, or how many points it ends up creating.

When things move to an irregular grid,
```
d = Interval(1.0, 10.0) #Continuous domain
grid_points = 1.0:0.01:10.0 #This is regular, but you get the point.
x = IrregularGrid(d, grid_points)
```
For the product of a discrete domain with 2 values and continuous domain (which comes up a lot in economics applications),
```
d = DiscreteDomain(2) * Interval(0.0,5.0) #Or tensor product?
```
Or, when we later want to enable naming of types for the discrete domain (e.g. for a continuous-time markov chain of switching between `L` and `H` productivity in a model)
```
d1 = Interval(1.0, 2.0)
d2 = DiscreteDomain([:L, :H])
d = d1 ⊗ d2 #Is the tensor product the right one to use here?
```
Or, instead, does it make sense to use an enumeration?  I really don't like how scoping works with enumerations in Julia, which limits the clarity of the code.

**Comment by ChrisRackauckas** [comment]
To get @MSeeker1340 up to speed, this is for being able to define the operators without the discretization. DiffEqOperators is all about how do you build the computational operator to compute, but there's a higher level question we should be asking as to, how can I just symbolically/mathematically specify some differential operator and have the correct computational operators be generated? At this level, it's not a finite difference, FEM, spectral, etc. discretization: just some operator on a domain which would make it easy to switch between forms (if we get it right).

ModelingToolkit.jl is an intermediate representation (IR) for models. You can think of it almost as a mini-CAS that DSLs can write into and compute on. The variables of this IR can hold domains (https://github.com/JuliaDiffEq/ModelingToolkit.jl/blob/master/src/variables.jl#L10) and there are differential operators. We can modify it at will to make the IR have what we need to generate operators. 

However, I think we should get some operators working first before we talk about auto-generating. In some sense, this is the "FEniCS" part and we need dolphin first.

**Comment by ChrisRackauckas** [comment]
>Maybe we can get some inspiration from Mathematica?

We're building this with help from REDUCE. 

>But yes I agree we should stick to simple finite-difference realizations of the operators at the moment, while maybe think a bit about incorporating ApproxFun.

Yes, first things first.

**Comment by dlfivefifty** [comment]
The tree-vs-type hierarchy question is important to get right. In ApproxFun I  overused templated variables, which puts a limit on how complicated the operator tree can be. However, building operators should be fast. It’s not clear to me what the right balance is.

**Comment by ChrisRackauckas** [comment]
>The tree-vs-type hierarchy question is important to get right. In ApproxFun I overused templated variables, which puts a limit on how complicated the operator tree can be. However, building operators should be fast. It’s not clear to me what the right balance is.

We did it a little wrong once. https://github.com/JuliaDiffEq/ModelingToolkit.jl/issues/62 . I think the key is to separate "compile time" and "runtime" here. The resulting operator can be fully typed, just the computation to get it may not be inferred. That's fine because building the operator would either be at the top scope or above a function barrier, i.e. building it is a form of compilation so like how the Julia AST itself is untyped, the specification of the operator should be fully untyped but builds strongly typed code.
We'll at least give it a try.

**Comment by jlperla** [comment]
@dlfivefifty @ChrisRackauckas  What are the thoughts on using `Interval(...)` here, exactly as in `ApproxFun`? vs. `0.0:1.0` vs. `0.0 .. 1.0`?

I was experimenting with instead using the return from the range built into Julia (i.e. `1.0:10.0`) but
```
julia> 1.0:10.0 |> typeof
StepRangeLen{Float64,Base.TwicePrecision{Float64},Base.TwicePrecision{Float64}}
julia> 1.0:10.0
1.0:1.0:10.0
```
As much as it would be great in notation to support defining intervals with `0.0:1.0`, I do not think this usage scenario fits the Range concept, which is inherently iterable.

The better fit is probably https://github.com/JuliaIntervals/IntervalRootFinding.jl  What about using the `..` operator?:
```
julia> using IntervalArithmetic

julia> 1.0..10.0 |> typeof
IntervalArithmetic.Interval{Float64}

julia> 1.0..10.0
[1, 10]
```
Are there other precedents for using this notation for intervals?

**Comment by dlfivefifty** [comment]
` a .. b `  is the agreed upon convention for intervals, see IntervalSets.jl

Abusing a ` UnitRange ` is a bad idea.

**Comment by dlfivefifty** [comment]
(It looks like you aren’t aware that ApproxFun also supports ` a .. b ` )

**Comment by jlperla** [comment]
Perfect!  `a .. b` as sugar to create the Interval it is.  Though this tells me that ApproxFun and IntervalArithmetic are incompatible, and explains the error I saw.  Should some sort of shared interval base library be established?

(Even in the longrun, with my ADL suggestion in https://discourse.julialang.org/t/function-name-conflict-adl-function-merging/10335 wouldn't help here, since the operator `..` needs to work on two floats in both cases)

For the grids, are there any other libraries that would be nice to maintain compatibility with?  For https://github.com/JuliaMath/Interpolations.jl  it seems that they use the `StepRangeLen` type directly for the grid, but I wonder if we are better off maintaining a connection from a grid to a domain.

**Comment by jlperla** [comment]
Also https://github.com/sisl/GridInterpolations.jl has its own grid objects

**Comment by dlfivefifty** [comment]
The whole point of IntervalSets.jl is to be a shared package for ` a .. b ` .

**Comment by jlperla** [comment]
Done.  What about grids?  Any movement towards a coordinated library for those types?

**Comment by dlfivefifty** [comment]
Not on METADATA but we’ve been working on https://github.com/daanhb/Domains.jl

Unfortunately it was triggering a compile-time Julia bug when used in ApproxFun, which put a hold on development. We also decided  “domain” should be an interface rather than an abstract type, and anything that supports ` in ` can be treated as a domain.

**Comment by jlperla** [comment]
@dlfivefifty What are your thoughts on domains with discrete numbers of elements?  Did you have that in mind with Domains.jl?  In particular, having them "named"?  I don't think using a enumeration is a good idea due to namespace clashes, but symbols work well.

@ChrisRackauckas Thoughts on getting uniform grids up and running for the prototype in DiffEqOperators.jl ?  I think that `IntervalSets.jl` as a dependency is an easy addition?  A temporary type to hold grids may be reasonable until things we are able to coordinate on a central grid type.

**Comment by dlfivefifty** [comment]
Making domain an interface has the benefit that, since ` 1 in 1 ` works, numbers can be interpreted as points and vectors can be interpreted as grids. So in 1D I’d just use ` linspace `.  For 2D one would want an array of ` SVector ` I think.

**Comment by jlperla** [comment]
What is the connection between domains, intervals, etc.?  It sounds like you are saying that the domain is the discretized grid that the function is defined on.  Is that correct?

If so, then what do we need for the generic algorithms to work:
- `in` is necessary, as you say
- Some testing of equality is needed?
    - The reason is ultimately to be able to ensure that operators we are composing are defined on the same domains, and potentially expand to the superset of the domain
- Don't we need iteration as well for generic algorithms to be implemented on the grids themselves?  If so, then https://docs.julialang.org/en/stable/manual/interfaces/#man-interface-iteration-1 would also be a requirement of a domain?

Where I start to get confused in the interface, is how this would all work for algorithms with multiple dimensions.  I see how the `in` works, but how do you write generic code?  I think this becomes especially important with designing the operator interface, because I presume the iteration is used to go through the state with the `update_coefficients!` design.

**Comment by dlfivefifty** [comment]
In my usage domain is a possibly infinite set. So all of these would implement the domain interface:

```
a..b 
a:b
1
[1,2,3]
Set([1,2,3])
```
The first is infinite (an interval) the rest are discrete.

Iteration is not possible with infinite sets but is for any of the discrete ones.

To generalise to higher dimensions, I interpret the “ambient space” of the domain as all elements of the type, and would use something like ` SVector{2,T} ` to represent 2D points.

@daanhb May have some thoughts.

**Comment by jlperla** [comment]
OK, then for the "named" discrete sets, vectors of symbols would work perfectly.
```
[:L, :H]
```

Thinking about your domain, the generality makes complete sense for spectral methods.  I guess the question with finite-difference methods is whether there is need to have separate generic types to for the continuous domain and the discretization of the continuous domain.  That was the assumption I had with my suggested
```
N = 20
d = 1.0..10.0
x = RegularGrid(d, N)
```
But, is the `d` really needed to keep around, or is the `x` enough?  For example, we could just go
```
x = 1.0:0.1:10.0 #For a regular grid
x = linspace(1.0, 10.0, 20) #Generates the same type
x_irregular = [1.0, 1.1, 1.5, 2.0, 5.0, 10.0] #Irregular grid
```
One worry in that approach would be whether you can have specialized code for regular vs. irregular grids.  But I think it works completely fine, since the type of the range and the vector are different and it could dispatch on `StepRangeLen`.  That is
`
julia> x = linspace(0.0, 1.0, 10)
0.0:0.1111111111111111:1.0

julia> typeof(x) <: StepRangeLen
true

julia> typeof(collect(x)) <: StepRangeLen
false

julia> step(x)
0.1111111111111111

`

**Comment by jlperla** [comment]
> We're only talking about grids now, continuous domains are for a higher level.
OK, then no need to have the separation of the continuous domain and the discretization of it.

Some specific quetsion: 
- So, lets say we are composing some operators.  Would we then have then keep the "grid" with each of them (which could be used to ensure that the composed operators are defined on the same grid type).  That is,
```
x = linspace(0.0, 1.0, 100) #100 node grid
sigma_2 = 1.1
mu = -0.2
L = mu * DriftOperator(x, :forwards) + sigma/2 * DiffusionOperator(x)
```
But what if there are the functions for the coefficients?  Is the idea that the "functions" would conform to whatever the "grid" was for the 

Would we

**Comment by jlperla** [comment]
Although I would love it if for multiple dimensions we could use a named tuple.  In particular, the canonical case in economics is a discrete `i in {L,H}` state which determines the drift of a continuous state, `y`.  For example, if the drift is `0.1 * y` if `i = :L` and `0.5 * y` if `i=:H`, I would love to be able to write this as something like:
```
julia> x = @NT(y=0.1, i=:L)
(y = 0.1, i = :L)

julia> μ(u,p,t,x) = x[:i]==:L ? 0.1*x[:y] : 0.5*x[:y]
μ (generic function with 1 method)

julia> μ(0,0,0,x)
0.010000000000000002

julia> μ(0,0,0,@NT(y=0.5, i=:H))
0.25
```
(where the `@NT` is obviously dropped in v0.7

**Comment by ChrisRackauckas** [comment]
>Although I would love it if for multiple dimensions we could use a named tuple. In particular, the canonical case in economics is a discrete i in {L,H} state which determines the drift of a continuous state, y. For example, if the drift is 0.1 * y if i = :L and 0.5 * y if i=:H, I would love to be able to write this as something like:

NamedTuples satisfy the interface. You can do that if you want.

>But what if there are the functions for the coefficients? Is the idea that the "functions" would conform to whatever the "grid"?

Yes, it would just be on the user side the `x` is of the right dimension. It would be like an FEM node listing.

**Comment by jlperla** [comment]
> NamedTuples satisfy the interface. You can do that if you want.

Is it that simple?  Presumably the interface of what is passed into the `x` of the function has something to do with the types of the dimensions of the grid?  Forgetting about the named tuples for a second, it seems like in the construction of the tuple itself the underlying code in the library is going to have to take the Cartesian product of all of the elements of the types?

It seems like if I want to have an operator which works on the product of a continuous and discrete grid, I would have to put it in a tuple, or something like that:
```
domain = ( linspace(0.0, 1.0, 10), [:L, :H] ) #A tuple of the domain
```
...or add syntactic sugar for the cartesian product.  Then generic code calling the functions in the library could construct a tuple to call the underlying functions of type `Tuple{eltype(domain[1]), eltype(domain[2])}`


And in order to have names, wouldn't we need to have those named associated with the  underlying domain.  That is, something like `AxisArrays`?
```
i_domain = Axis{:i}([:L, :H])
y_domain= Axis{:y}(1.0:0.1:2.0)
```
That way, a named tuple could be created to call the user functions by getting the names of each axis with `AxisArrays.axisname(i_domain)` etc. and then the type with `eltype(i_domain)`, etc.

**Comment by jlperla** [comment]
OK, but
```
julia> x = collect(linspace(0.0, 1.0, 5))
5-element Array{Float64,1}:
 0.0
 0.25
 0.5
 0.75
 1.0

julia> y = [:L, :H]
2-element Array{Symbol,1}:
 :L
 :H

julia> z = [1, 2]
2-element Array{Int64,1}:
 1
 2

julia> SVector.(x, z')
5×2 Array{StaticArrays.SArray{Tuple{2},Float64,1,2},2}:
 [0.0, 1.0]   [0.0, 2.0]
 [0.25, 1.0]  [0.25, 2.0]
 [0.5, 1.0]   [0.5, 2.0]
 [0.75, 1.0]  [0.75, 2.0]
 [1.0, 1.0]   [1.0, 2.0]

julia> SVector.(x, y')
ERROR: MethodError: no method matching transpose(::Symbol)
```
Basically, this approach of the cartesian products seems to promote discrete dimensions, and doesn't allow symbols?

If I add in a transpose,
```
julia> Base.transpose(x::Symbol) = x

julia> SVector.(x, y')
5×2 Array{StaticArrays.SArray{Tuple{2},Any,1,2},2}:
 Any[0.0, :L]   Any[0.0, :H]
 Any[0.25, :L]  Any[0.25, :H]
 Any[0.5, :L]   Any[0.5, :H]
 Any[0.75, :L]  Any[0.75, :H]
 Any[1.0, :L]   Any[1.0, :H]
```
It looks the "promotion" turns these into "Any", which is not what we want.

Isn't the basic problem that `SArray` are designed to hold a single type?

**Comment by dlfivefifty** [comment]
Then use ` tuple `. ` SVector ` is only needed if you want to add your points.

**Comment by jlperla** [comment]
OK.  So
```
julia> using IterTools

julia> grids = (1.0:0.1:2.0, [:L, :H])
(1.0:0.1:2.0, Symbol[:L, :H])

julia> [i for i in product(grids...)]
22-element Array{Tuple{Float64,Symbol},1}:
 (1.0, :L)
 (1.1, :L)
 (1.2, :L)
 (1.3, :L)
...
```
So at least `IterTools` could be used to get an enumeration on the cartesian problem?  I suppose something similar with a `Axis` generating a `NamedTuple` would also be possible?

**Comment by ChrisRackauckas** [comment]
Yes, any of these `AbstractVector` with scalars or `AbstractVector`s are a good way to specify the points of a domain. It doesn't matter which you choose, though being able to determine `dx` will matter for operators.

**Comment by ChrisRackauckas** [comment]
>Although in the case of discrete states, the dx is not meaningful, since the infinitesimal generator is intensity matrix describing the jump intensity of the continuous-time Markov-chain.

Yes exactly, so it would need an array operator and this space wouldn't be compatible with most of the operators. 

>A completely different alternative is that we don't generally let the user define functions of everything, and instead compose up the operators themselves by taking kroneckers of the underlying operators on univariate operators.

Yup that's what we're doing.

**Comment by jlperla** [comment]
@ChrisRackauckas @MSeeker1340 @dlfivefifty 
I wanted to resurrect this issue since we are making a push on adding differential operators.

I need to have both regular and irregular grids in the very short term, and an irregular grid X a discrete state in the short/medium term, so I think it is worth designing the interface correctly.

My general feeling is that any derivative operator should you should always pass in the "grid", where if it is an `AbstractRange` you can call the `step(r)` to get the size of the delta for uniform grids, and if it is `AbstractArray` then you can assume that it could be irregular.  Later on, we could check that the grids `===` when we compose different operators, or do some fanciness to take the union of them if they don't.

That is, something like
```julia
x = linspace(0.0, 1.0, 10) #regular grid since returns an AbstractRange
L = mu * DriftOperator(x, :forwards) +sigma_2 * DiffusionOperator(x)

x2 = collect(x) #Or create an irregular grid
L2 = mu * DriftOperator(x2 , :forwards) +sigma_2 * DiffusionOperator(x2 ) #Same basic interface
```
Alternatively, maybe it makes sense to separate the interface for the grid and the underlying domain/space, etc.  but that is a design question above my paygrade for integrating ApproxFun and the DiffEqVerse.

The interface for taking the 2 dimension grid of a discretized continuous state X an enumerable discrete state is the other place that the design of the grids/domains/etc. enters.  But lets say I want to have a PDE in a continuous `x`  variable and a discrete variable `i` in `:l` and `:h` where there are different drifts in the stochastic process for the `:L` vs `:H`.  Then I would love to be able to do something like
```julia
x = linspace(0.0, 1.0, 10) 
D = DriftOperator(x, :forwards) 
D2 = DiffusionOperator(x)
sigma = 0.01
mu_L = 0.02
mu_H = 0.05

L_L = mu_L * D  + sigma^2/2 * D2
L_H = mu_H * D + sigma^2/2 * D2
#Stack the independent operators?  Making this up
L = [:L 0; 0 :H] * [L_L; L_H]; %Product of discrete state and continuous state?
```
Obviously the last state is made-up, but I would how we could have something of similar complexity.

**Comment by dlfivefifty** [comment]
I don’t remember what the plan was but I’m pretty sure the right design is a software tiers that “lower”. That way ApproxFun+DiffEq can use the appropriate tiers while not forced to integrate everything on the operator level.

So For finite differences we want the following, from highest to lowest:

1. Differential operators and boundary conditions (like you wrote)
2. Lazy banded matrices and lazy constraints
3. Concrete banded matrices (or SymTridiagonal when possible) with concrete constraints.

 The linear algebra happens to incorporate constraints is either at tier 2 or 3.

ApproxFun (without adaptivity) has an extra step:
1. Differential operators and boundary conditions
2. Lazy infinite banded matrices with constraints 
3. Lazy finite banded matrices with constraints 
4. Concrete almost banded matrices 

In the future, there could be another layer on top that represents the differential operators without any discretisation details, that can then be solved via either approach.

The reason for the extra “lazy” tier is that this is going to make distributed memory super easy, as the lazy matrices will be what’s communicated.

**Comment by jlperla** [comment]
For the creation of the `Q` boundary extrapolation (if you recall the setup documented in the latex) could the notation be (using https://github.com/JuliaApproximation/ApproxFun.jl/issues/590).

Boundary condition for `u'(0) + a * u(0) = 0` and `u'(xbar) = 0` be something like

```
Q = [(𝒟 + a * I )→ 0; 𝒟 → xbar]
```
Then the composed operator applying the boundary conditions would be 
```
x = linspace(0.0, 1.0, 10) #regular grid since returns an AbstractRange
L1 = mu * DriftOperator(x, :forwards) +sigma_2 * DiffusionOperator(x)
L = Q * L1 #Applies boundary extrapolation to L1
```
Where at that point we can use `L` as a linear map (or affine, in general)?

ApproxFun.jl, as far as I understand it, does not need to separate boundary conditions from equations, so the other approach is simply
```
x = linspace(0.0, 1.0, 10) #regular grid since returns an AbstractRange
L1 = mu * DriftOperator(x, :forwards) +sigma_2 * DiffusionOperator(x)
L = [L1;
      (𝒟 + a * I )→ 0;
      𝒟 → xbar];
```
But I am not sure how that works with the `Q` approach here.  These sort of things are sugar we can wait on, but thinking through the interface might be worthwhile.

**Comment by dlfivefifty** [comment]
ApproxFun doesn't support time-stepping, though it supports time-evolution linear PDEs like zero Dirichlet Heat equation a la (using the to-be-implemented new syntax for restriction operators, and using `u(t,x)`):
```julia
Dx = Derivative([0,1])
Dt = Derivative([1,0])
[ I → (0..1) × [-1,1];
  I → 0 × (0..1);
  Dt - Dx^2 ]  \ [[0,0], u_0, 0]
```
In theory, `newton` and it's auto-diff parser will eventually work for non-linear PDEs, here would be Burgers:
```julia
    N=u->[u(:,-1);
                u(:,1);
                u(0,:)-u_0;
                 Dt*u + u*Dx*u - ν*Dx^2*u]
    u = newton(N) # sequence of linear solves
```

In terms of time-stepping and DiffEq, I'm imagining something like (for Dirichlet heat, copying DAE example):
```julia
function heat(out, du, u, _, t)  
    out[1] = 𝒟^2*u - du
   out[2] = ( I → -1) * u
   out[3] = ( I →   1) * u
end
prob = TimeEvolutionPDEProblem(heat, u_0, tspan)
solve(prob, FiniteDifferences(1000)) # Or Ultrapherical() or Fourier() for ultraspherical spectral method a la ApproxFun
```
and for Burgers
```julia
function burgers(out, du, u, ν, t)  
    out[1] = du  + u*(𝒟*u) - ν*𝒟^2*u
   out[2] = ( I → -1) * u
   out[3] = ( I →   1) * u
end
prob = TimeEvolutionPDEProblem(burgers, u_0, tspan)
solve(prob, FiniteDifferences(1000)) # Or Ultraspherical() or Fourier() for ultraspherical spectral method a la ApproxFun
```

**Comment by dlfivefifty** [comment]
I don't understand the `Q * L` part.

I think we need to focus on linear for now, as nonlinear is more complicated (and may require some auto-diff to get working).

Instead of worrying about PDEs first, we should get the lowest level right: which might be some sort of `ConstrainedLinearODEProblem` that involves only matrices. So maybe something like this would work:
```julia
Δ = BandedMatrix{Float64}(undef, (n-2, n), (0,2))
Δ[band(0)] = Δ[band(2)] = 1/h^2
Δ[band(1)] = -2/h^2

B = zeros(2,n)
B[1,1] = B[end,end] = 1

R = eye(n)[2:end-1,:]

prob = ConstrainedLinearODEProblem(B, R, Δ)

solve(prob, GaussReduce(Tsit5(), pivots = (1,n)))
```
Note the pivots can be any two columns of `B` that are non-zero, and could also be calculated on the fly.

This is too simple as nothing varies in time. If we have a simple time-varying PDE like `u_t = Δ*u + cos(t)*u` it could be something like:
```
prob = ConstrainedLinearODEProblem(t -> B, t -> R, t -> Δ + cos(t)*I)

solve(prob, GaussReduce(Tsit5(), pivots = (1,n)))
```
Not sure what to do if the pivots themselves need to vary (think `cos(t)*Dirichlet + sin(t)*Neumann`)...

This is also too simple as there is no support for forcing terms. Something like:
```
prob = ConstrainedLinearODEProblem(t -> (B,c), t -> R, t -> (Δ + cos(t)*I,f))

solve(prob, GaussReduce(Tsit5(), pivots = (1,n)))
```

Definitely needs some iteration....

**Comment by ChrisRackauckas** [comment]
@dlfivefifty 's syntax up there, at least the first one, has far too much magic and only supports DAEs so I don't like that one. I would like to keep everything as short syntactic operators, or use a DSL. I think it should commit to being automatic or being fully flexible, since anything in between gets complicated but limited.

>I don't understand the Q * L part.

That's a generic way to handle linear boundary conditions on any linear operator. 

>I think we need to focus on linear for now, as nonlinear is more complicated (and may require some auto-diff to get working).

Nonlinear has been working for over a year now and with autodiff. 

>Instead of worrying about PDEs first, we should get the lowest level right: which might be some sort of ConstrainedLinearODEProblem that involves only matrices.

That's not necessary. A ConstrainedLinearODEProblem is just an ODEProblem with a linear operator and a mass matrix. 

Anyways, there's no reason to argue at that level of the design because that's very clear anyways. It's the discretization handling for named PDEs that's unclear.

**Comment by dlfivefifty** [comment]
The whole point is `ConstrainedLinearODEProblem` is _not_ the same as a mass matrix. One way to solve a `ConstrainedLinearODEProblem` is to reduce it to a mass matrix formulation by Gaussian elimination, but there exists other methods like IMEX which incorporate the constraints in the implicit solves.

(A special case is when `R` is a view of the identity matrix, in which case we can reduce further to a standard ODE problem.)

Mixing in the language of finite differences / differential operators / boundary conditions with the actual details of the solve, which only involves linear algebra on a constrained linear ODE, will be limiting.

**Comment by ChrisRackauckas** [comment]
>The whole point is ConstrainedLinearODEProblem is not the same as a mass matrix. One way to solve a ConstrainedLinearODEProblem is to reduce it to a mass matrix formulation by Gaussian elimination, but there exists other methods like IMEX which incorporate the constraints in the implicit solves.

Yes, and that's a SplitODEProblem with a mass matrix. That was exactly my point.

>Mixing in the language of finite differences / differential operators / boundary conditions with the actual details of the solve, which only involves linear algebra on a constrained linear ODE, will be limiting.

That's exactly why I don't want to do that.

**Comment by dlfivefifty** [comment]
It's a `SplitODEProblem` with a mass matrix (`R = view(Eye(n),2:n-1,:)` in the case of FD) and also boundary conditions. I don't see how boundary conditions are incorporated in a standard `SplitODEProblem` without further manipulation.

Also, we should have an idea of what to do with 2D time evolution. Here there's the added difficulty that boundary conditions over-pose the constraints (and work because the RHS is in a special subspace).

**Comment by ChrisRackauckas** [comment]
Boundary conditions are just handled like any other conditions by the mass matrix. 

>Also, we should have an idea of what to do with 2D time evolution. Here there's the added difficulty that boundary conditions over-pose the constraints (and work because the RHS is in a special subspace).

That's operator-specific. The finite difference ones can lazy Kron. Since the intermediate level isn't supported, as long as each operator should just do whatever is appropriate for it.

**Comment by dlfivefifty** [comment]
> Boundary conditions are just handled like any other conditions by the mass matrix.

But the user has to do this. I thought the whole point was to make a system that incorporates the boundary conditions automatically, without the user having to set-up their own mass matrices?

**Comment by dlfivefifty** [comment]
At the intermediate level there are no operators, just matrices. It's not a complex interface, it's the `AbstractArray` interface. And if nonlinearity can be incorporated at all, it can definitely be incorporated at this level.

By doing things on the level of linear algebra, you have something that other discretisation methods can also use.

**Comment by ChrisRackauckas** [comment]
I'll agree that level is nice in theory and can be useful for composing different pieces, but it's not a level that anyone has really expressed interest in using. The lower level methods are for implementation of discretizations so it's for the developers and those who really just want helpers for building their own solver. The high level methods are what people have expressed a lot of interest in: can I guess say in a high level way that I want you to solve a semilinear diffusion advection equation, and have you come up with how to do it correctly. The middle, expressing the abstract PDE with usable operators that double as code, is really cool and likely where I want to be in the long term but its a lot of work looking for a problem.

**Comment by dlfivefifty** [comment]
It's the level no one wants because they don't realise everything is just structured linear algebra. When it comes to parallelisation, working in this language will make parallelising a lot easier, because it can be done on the linear algebra level, not on the level of discretisation.

It sounds like you want to go down the same route as FeNICs, Firedrake, Dedalus, etc. which specialise on particular discretisation methods, and jumble up the linear algebra with the details of the discretisation, and with the details of parallelising.

 This is fine if you only ever want to solve PDEs, and specific ones at that. But I'm more interested a framework that is flexible enough to handle other problems just as easily (boundary element methods, singular integral equations, Benjamin–Ono equation, fractional differential equations).

**Comment by dlfivefifty** [comment]
Though maybe our wires are crossed. The really important thing is that `SplitODEProblem` supports lazy structured matrices (that may depend on time) that are realised as a distributed structured matrix remotedly: that is, the workers have access to the lazy matrix and know how to populate their blocks.

The mass matrix vs boundary conditions conversation is less crucial, and actually you might be right that its hard to incorporate boundary conditions automatically into the mass matrix while retaining structure.

**Comment by jlperla** [comment]
> I don't understand the Q * L part.

The `Q` stuff comes from the https://github.com/JuliaDiffEq/PDERoadmap/releases/download/v0.2/linear_operators_overview.pdf which in turn was formalizing the algebra for our discussion in https://github.com/JuliaDiffEq/DifferentialEquations.jl/issues/260

You can see the examples there.

But I also think that I did something confusing here.  You cannot create the `Q` directly from the `B` as it requires the `L`.  So perhaps something like
```julia
B = [( I → -1);  ( I →   1)];
L = 𝒟^2
Q = boundaryextrapolation(B, L)
heat = L * Q 
prob = ODEProblem(heat, u_0, tspan)
```
Probably better is if the function to just extrapolate to the boundary and create the internal `L * Q` is also defensible.  That could be done by overloading the vcat
```
M = 100
x = linspace(0.0, 1.0, M)
B = [( I → -1);  ( I →   1)];
L = 𝒟(x)^2
L_q = [L; B] #Interally creates a `L*Q` setup, creating the correct Q from L and B
prob = ODEProblem(L_q, u_0, tspan)
```
This relies on the `update_coefficients!` for the internal linear operators.

Or, better yet, since  `L_q : R^M -> R^M` and is linear, we can use an exponential integrator.  Or did I miss something here?

**Comment by dlfivefifty** [comment]
I think `[L; B]` is a bad idea: I used to do something similar for `eigs` in ApproxFun, and its too much of an abuse of notation. Also, how do you include the forcing terms? (for problems like `B*u = c` and `u_t = L*u + f`)

Has it come up that eigenvalue problems have the _exact_ same issue (and solution) of incorporating the boundary conditions into the "mass" matrix of a generalised eigenvalue problem? Starting with the syntax for `eigs` (and the implemention) might not be a bad idea as its simpler than time-stepping: no nonlinear terms. I would say for `eigs` the following syntax seems reasonable:
```julia
eigs(L; constraints=(B,c))
```

**Comment by jlperla** [comment]
Yeah, you are right, that doesn't work.  What I do know is that:
- You can't find the `Q` without knowing both the `L` and the `B` together.  In fact, you can't even know the sizes of the matrices involved until you know the stochastic process `L` (e.g. jump-diffusion boundary conditions have an extension of more just one ghost node)
- We don't know the general approach to calculating the `Q` from a `B` and an `L`.
- It is not really useful to have the `Q` on its own, as we only use the `L * Q`  product for calculations.  However, in most cases that interior of the `Q` is an identity function, so the `L * Q` product can probably be done in a smart and lazy way.
- We had originally wrote this to take a linear `L` and generate a potentially affine `L * Q` (i.e. it could include the forcing terms for the boundary values, but we didn't write it for forcing terms in the differential operator).  Should that be revisited?

Lets take this discussion to a new issue to focus on the interface, as opposed to the discussion of the domain here.

**Comment by ChrisRackauckas** [comment]
>We don't know the general approach to calculating the Q from a B and an L.

Yes we do. It's just writing down a choice of interpolating polynomial on the full space which satisfies the BCs.

>It is not really useful to have the Q on its own, as we only use the L * Q product for calculations. However, in most cases that interior of the Q is an identity function, so the L * Q product can probably be done in a smart and lazy way.

We can build it into the pre-made operators if we want to, and that would make them square. That's a lot like the current DiffEqOperators.jl design, but with a few fixes to some portions by standardizing on only doing the interior.

**Comment by jlperla** [comment]
> Yes we do. It's just writing down a choice of interpolating polynomial on the full space which satisfies the BCs.

Let me rephrase: at the point of us writing this up, and discussing with you, we did not have an algorithmic way to take a `B`, `L`, and spit out a `Q`.  Though we have ways to check if a particular `Q` is correct.  If you have an algorithm to create the `Q` then write it down so that @MSeeker1340 can consider implementing it.

But considering how tricky the generic machinery for this could be (i.e. we don't even know the size of the extension space until we compose all operators) I am not 100% sure it is necessary yet.  Adding to that the hope to make the `Q` part of the `L * Q` lazy, and it seems like we shouldn't go too general too quickly.

> We can build it into the pre-made operators if we want to, and that would make them square. That's a lot like the current DiffEqOperators.jl design, but with a few fixes to some portions by standardizing on only doing the interior.

For me, the more that this stuff simply does a more complete job of implementing DiffEqOperators.jl with operator composition, the happier I get.  While I like the idea of making the interface look nice and DSL like, I don't like the idea of radically increasing the complexity of what this is intended to do.

**Comment by dlfivefifty** [comment]
Here's code that spits out `Q` for any `B` and `L`, using just linear algebra (no interpolating polynomials):
```julia
using LinearAlgebra

n = 10
h = 1/n
L = zeros(n-2,n); L[diagind(L,0)] .= 1/h^2; L[diagind(L,2)] .= 1/h^2; L[diagind(L,1)] .= -2/h^2;

function projection(B, L)
    n = size(L,2)
    @assert size(B,2) == n
    @assert size(B,1) == 2 # can be generalised to arbitrary number of BCs

    j1 = 0
    for j=1:size(B,2)
        if norm(B[:,j]) ≠ 0
            j1 = j
            break
        end
    end
    j2 = 0
    for j=j1+1:size(B,2)
        if rank(B[:,[j1,j]]) == 2
            j2 = j
            break
        end
    end

    B = inv(B[:,[j1,j2]])*B

    # eliminate j1 & j2
    P = Matrix(I,n,n)[:,[j1,j2]]
    P̃ = Matrix(I,n,n)[:, setdiff(1:n, [j1,j2])]

    (I-P*B)*P̃
end

# dirichlet
B = zeros(2,n); B[1,1] = B[2,end] = 1
Q = projection(B,L)
L*Q
# 8×8 Array{Float64,2}:
#  -200.0   100.0     0.0     0.0     0.0     0.0     0.0     0.0
#   100.0  -200.0   100.0     0.0     0.0     0.0     0.0     0.0
#     0.0   100.0  -200.0   100.0     0.0     0.0     0.0     0.0
#     0.0     0.0   100.0  -200.0   100.0     0.0     0.0     0.0
#     0.0     0.0     0.0   100.0  -200.0   100.0     0.0     0.0
#     0.0     0.0     0.0     0.0   100.0  -200.0   100.0     0.0
#     0.0     0.0     0.0     0.0     0.0   100.0  -200.0   100.0
#     0.0     0.0     0.0     0.0     0.0     0.0   100.0  -200.0
#


# neumann
B = zeros(2,n); B[1,1] = -1; B[1,2] = 1; B[2, end-1] = -1; B[2,end] = 1
Q = projection(B,L)
L*Q
# 8×8 Array{Float64,2}:
#  -100.0   100.0     0.0     0.0     0.0     0.0     0.0     0.0
#   100.0  -200.0   100.0     0.0     0.0     0.0     0.0     0.0
#     0.0   100.0  -200.0   100.0     0.0     0.0     0.0     0.0
#     0.0     0.0   100.0  -200.0   100.0     0.0     0.0     0.0
#     0.0     0.0     0.0   100.0  -200.0   100.0     0.0     0.0
#     0.0     0.0     0.0     0.0   100.0  -200.0   100.0     0.0
#     0.0     0.0     0.0     0.0     0.0   100.0  -200.0   100.0
#     0.0     0.0     0.0     0.0     0.0     0.0   100.0  -100.0

# IVP
B = zeros(2,n); B[1,1] = B[2,2] = 1
Q = projection(B,L)
L*Q

# 8×8 Array{Float64,2}:
#   100.0     0.0     0.0     0.0     0.0     0.0     0.0    0.0
#  -200.0   100.0     0.0     0.0     0.0     0.0     0.0    0.0
#   100.0  -200.0   100.0     0.0     0.0     0.0     0.0    0.0
#     0.0   100.0  -200.0   100.0     0.0     0.0     0.0    0.0
#     0.0     0.0   100.0  -200.0   100.0     0.0     0.0    0.0
#     0.0     0.0     0.0   100.0  -200.0   100.0     0.0    0.0
#     0.0     0.0     0.0     0.0   100.0  -200.0   100.0    0.0
#     0.0     0.0     0.0     0.0     0.0   100.0  -200.0  100.0
```

**Comment by jlperla** [comment]
Great! Can you write a version of this that works for a general setup with more than one extension node in each direction? That is, let there me a total of `M_E_m` on the bottom, `M_E_p` on the top so that `M_bar = M + M_E_m + M_E_p` where the B has M_bar columns, etc.

**Comment by dlfivefifty** [comment]
I don't know what you mean "top" and "bottom". Do you mean 3 boundary conditions, 2 at the left and 1 at the right? Here it is:
```julia

function projection(B, L)
    n = size(L,2)
    b = size(B,1)
    @assert size(B,2) == n
    @assert size(L,1) == n - b

    js = Vector{Int}()
    for ξ = 1:b
        j1 = 0
        for j=1:size(B,2)
            if rank(B[:,vcat(j,js...)]) == ξ
                j1 = j
                break
            end
        end
        push!(js,j1)
    end
    @show js
    B = inv(B[:,js])*B

    # eliminate j1 & j2
    P = Matrix(I,n,n)[:,js]
    P̃ = Matrix(I,n,n)[:, setdiff(1:n, js)]

    (I-P*B)*P̃
end

# linear kdv u_t = u_xxx
n = 10
h = 1/n
L = zeros(n-2,n); L[diagind(L,0)] .= 1/h^2; L[diagind(L,2)] .= 1/h^2; L[diagind(L,1)] .= -2/h^2;
D = zeros(n-3,n-2); D[diagind(D,0)] .= -1/h; D[diagind(D,1)] .= 1/h;
L = D*L

# IVP + Right neumann
B = zeros(3,n); B[1,1] = 1; B[2,1] = -1; B[2,2] = 1; B[3,end-1] = -1; B[3,end] = 1
Q = projection(B,L)
L*Q
# 7×7 Array{Float64,2}:
#  -3000.0   1000.0      0.0  …      0.0      0.0      0.0
#   3000.0  -3000.0   1000.0         0.0      0.0      0.0
#  -1000.0   3000.0  -3000.0         0.0      0.0      0.0
#      0.0  -1000.0   3000.0      1000.0      0.0      0.0
#      0.0      0.0  -1000.0     -3000.0   1000.0      0.0
#      0.0      0.0      0.0  …   3000.0  -3000.0   1000.0
#      0.0      0.0      0.0     -1000.0   3000.0  -2000.0
```

**Comment by dlfivefifty** [comment]
Though I'm really confused why we needed `Q` in the first place... What's wrong with using the fact that
```julia
B u = c    <=> B_t u + B*u_t = 0
```
where `B_t` is the derivative of `B` w.r.t. `t` (which is just zero for Dirichlet/Neumann) to write the ODE as 
```julia
[B; R]*u_t = [B_t; L]*u
```
Then to get a regular ODE we have:
```julia
u_t = inv([B; R])*[B_t; L]*u
```

This simple approach seems to also recover the standard Dirichlet/Neumann matrices.

**Comment by jlperla** [comment]
> Do you mean 3 boundary conditions, 2 at the left and 1 at the right?

Yeah, I guess I am having a little trouble separating the count of what we might call the equations for the boundary conditions from the number of discretized equations that it generates. The key to me is that with jumps the size of the extended space can get bigger than just adding two points when discretized. Typically you would write the boundary conditions as a single equation (pre discretization)... For example a reflection of the jump... But I think that adds in multiple boundary condition equations when discretized

**Comment by dlfivefifty** [comment]
Can you write down the differential equation with boundary conditions you had in mind?

**Comment by ChrisRackauckas** [comment]
>Though I'm really confused why we needed Q in the first place... What's wrong with using the fact that

That's a mathematical solution, but not a computational one. `inv([B; R])` is not a nice operation, while the matrix that it expands to is simple and easy to just directly write down via an alternative derivation by just extrapolating from the interior to the boundary with the right order. It's mathematically equivalent but it's the difference between requiring large dense operators or a matrix-free lazy operator.

**Comment by dlfivefifty** [comment]
OK, in this case we can write 
```julia
[B[1,:]; R; B[2,:]] = I + U*V'
```
where `U` is sparse and use the Woodbury formula
```julia
inv([B[1,:]; R; B[2,:]]) = I - U*inv(I + V*U)*V
```
and this will still be sparse. But I think it's just writing the same thing in different ways.

**Comment by wormell** [comment]
It may not parallelise well but one could also view the problem as 
'''julia
u_t = [B; R] \ ([B_t; L]*u)
'''
which is a sparse problem.

**Comment by ChrisRackauckas** [comment]
Sparse is still much much larger than O(1).

**Comment by dlfivefifty** [comment]
I think I understand the issue now: any differential operator can be thought of as either left or right derivatives. This doesn't actually change the definition of `D`, but it does change the definition of `R`, so let me . Here's some examples:
```julia
n = 10;
h = 1/n
# is either left or right differential operator
D = zeros(n-1,n); D[diagind(D,0)] .= -1/h; D[diagind(D,1)] .= 1/h;
# R is restriction for right differential operator, that is, from x_0,…,x_n to x_1,…,x_n
R = function(n)
    R = zeros(n-1,n); R[diagind(R,1)] .=1 ; R
end
# L is restriction for left differential operator, that is, from x_0,…,x_n to x_0,…,x_{n-1}
L = function(n)
    L = zeros(n-1,n); L[diagind(L,0)] .=1 ; L
end



# we need to choose R/L so that B is zero in the corresponding columns
# we can find the correct columns automatically
function nonsingularcolumns(B)
    js = Vector{Int}()
    b = size(B,1)
    for ξ = 1:b
        j1 = 0
        for j=1:size(B,2)
            if rank(B[:,vcat(j,js...)]) == ξ
                j1 = j
                break
            end
        end
        push!(js,j1)
    end
    js
end


# A restriction is a combination of Rs and Ls
function restriction(num_R, num_L, n)
    C = Matrix(I,n,n)
    for k = 1:num_R
        C = R(n-k+1)*C
    end
    for k = 1:num_L
        C = L(n-num_R-k+1)*C
    end
    C
end
    
# to determine number of R's and L's, we need to see how many "left" and how many "right"
# boundary conditions there are
function restriction(B)
    n = size(B,2)
    js = nonsingularcolumns(B)

    num_R = sum(js .< n/2)
    num_L = sum(js .> n/2)
    restriction(num_R, num_L, n)
end

# Dirichlet
B = zeros(2,n); B[1,1] = B[end,end] = 1
restriction(B)
# 8×10 Array{Float64,2}:
#  0.0  1.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0
#  0.0  0.0  1.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0
#  0.0  0.0  0.0  1.0  0.0  0.0  0.0  0.0  0.0  0.0
#  0.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0  0.0  0.0
#  0.0  0.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0  0.0
#  0.0  0.0  0.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0
#  0.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0  0.0  0.0
#  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0  0.0

# IVP
B = zeros(2,n); B[1,1] = 1; B[2,1] = -1; B[2,2] = 1;
restriction(B)
# 8×10 Array{Float64,2}:
#  0.0  0.0  1.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0
#  0.0  0.0  0.0  1.0  0.0  0.0  0.0  0.0  0.0  0.0
#  0.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0  0.0  0.0
#  0.0  0.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0  0.0
#  0.0  0.0  0.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0
#  0.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0  0.0  0.0
#  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0  0.0
#  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0


# KDV IVP + Right BC

B = zeros(3,n); B[1,1] = 1; B[2,1] = -1; B[2,2] = 1; B[3,end] = 1;
restriction(B)

# 7×10 Array{Float64,2}:
#  0.0  0.0  1.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0
#  0.0  0.0  0.0  1.0  0.0  0.0  0.0  0.0  0.0  0.0
#  0.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0  0.0  0.0
#  0.0  0.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0  0.0
#  0.0  0.0  0.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0
#  0.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0  0.0  0.0
#  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0  0.0
```

**Comment by jlperla** [comment]
> Can you write down the differential equation with boundary conditions you had in mind?

Sorry for the long response, but let me try to give a scetch of it.  I am having trouble figuring out a few of the detailed boundary conditions as finance people tend to be sloppy when defining them, but here is the basic setup.
- There is a stochastic process, `X_t` which is the payoff of an asset.  Let the stochastic process for this process have a linear infinitesimal generator `L_s`.
- The payoffs are discounted at interest rate `r`.  Then the HJBE for the function `u(x)` which values the stream of assets is (forgetting about boundary conditions for a second)
```
r u(t, x) = x + L_s u(t, x) + du(t,x)/dt
```
If `L_s` was a diffusion, then this would be a Kolgorov **Backwards** Equation.  Assume that we are looking at the stationary solution so that `du/dt = 0`.  Then the bellman equation can be written as
```
L u(x) = x
```
where `L := r - L_s` or we could write `L := r * I - L_s`

OK, lets talk about discretizing the `L` for a particular process before we talk about the boundary conditions.  Define the `x` on a grid of `M` points, and let `dx=1` just to be make things easier.  A big question is what the size of the extension is, i.e. `M_bar`.

__Warmup with processes we already know:__
- If `X_t = a dt` for `a < 0` then we know we should use backward differences, and the size of the extension must be (at least) `M_bar = M + 1` for the left ghost-node.   So the discertized `L_s` is of size `M X (M+1)`.   The stencil is left for the reader :-) , but you could obviously pad an additional column of 0s to get it to `M X (M+2)` if you wanted to.  If `a > 0` and you wanted to use forward differences, then it is the same story except the ghost node is "on the right".
- If `X_t = c W_t` for brownian motion, then there is a ghost node on both sides, and `L_s` is of size `M X (M+2)`
- Obviously, when composing the `r` component of the `L := r - L_s` composed process, the size without any fudges is just a `M X M` but we can always stack columns of zeros on left and right to compose with the others.

__Now for the trickier example:__
Lets say that the stochastic process for `X_t` is a jump diffusion.  For it we will compose `L_s = a L_1 + sigma^2/2 L_2 + lambda L_J`  where `a < 0`,  `L_1` is backwards first differences, `sigma` is the variance of the diffusion process, `L_2` is the stencil for central differences for the second differences, `lambda` is the poisson arrival rate of the jump process, and `L_J` is the generator of the jump process.

So what is the size of `M_bar`?  We don't know yet, but we do know it is at least `M + 2` from the `L_2` one.

As an example jump process, assume that things are such that upon an arrival, there is a 50% chance you jump forward by 2 and 50% chance you jump backwards by 3.  Since we assumed that `dx=1` in the grid, this means we can grid points back or two forward.  Let `M=3` then I am pretty sure that the generator is of size `(3+5)X M` and I think it is
```
L_J = [.5 0  0 -1  0 .5  0  0
       0 .5  0  0 -1  0 .5  0
       0  0 .5  0  0 -1  0 .5]
```
Finally, to compose the whole jump diffusio, now that we know the size is `M + 5` we have to pad the left and right side.  Some while the original first differences was something like
```
L_1 = [1 -1  0  0
       0  1 -1  0
       0  0  1 -1]
```
we can pad as
```
L_1^ex = [zeros(3,2) L_1 zeros(3,2)]
```
and similarly
```
L_2 = 0.5*[1 -2  1  0  0
           0  1 -2  1  0
           0  0  1 -2  1]
```
and 
```
L_2^ex = [zeros(3,2) L_1 zeros(3,1)]
```
.... now I could have made a mistake somewhere in all of this, but I hope you get the point.

As for the boundary conditions, the ones I am interested in are linear and affine but hopefully this gets my point across on why composition and boundary conditions is tricky.  As for the number, we need to pin down 5 numbers.

One example of a boundary condition is a reflecting barrier. This is `u'(x_bar) = 0` for a diffusion process, but I am not 100% sure how to write it for a jump-diffusion.  But if there was a reflecting barrier at the top then I think that this gives two equations and they are something like the following `B`:
```
[0 0 0 0 0 0  1 -1
 0 0 0 0 0 1 -1  0]
```

The other, trickier, boundary conditions are absorbing ones that are affine.  The standard setup in finance and economics is that there is a stopping point at the boundary with a particular nonzero payoff.  I think what you need to do there is define the payoffs crossing the boundary.  So if our grid was `x = [0 1 2]` with the `dx` and the 3 jumps back, 2 jumps forward, then we need to define a payoff at `x = [-3 -2 -1]` given the 3 jumps backwards.  These would give our 3 final equations for the setup.  I am trying to write that boundary condition formally, but it will take time.

CAVEAT: I am pretty sure I made errors in the actual matrices, but I think I have all of the sizes correct.

**Comment by dlfivefifty** [comment]
I'm confused by your explanation: my understanding is that one typically starts with a stochastic process, generates a continuous Fokker Plank PDE, then discretises the PDE with finite differences. It sounds like you want to jump directly from the stochastic process to the finite differences, and skip the continuous PDE? 

In terms of the continuous PDE, the description of the jump process doesn't sound right, as it's described in terms of jumps on the (artificial) grid.

I also don't understand "reflecting": this makes sense for wave equations and is equivalent to Neumann, but when you don't have waves, what does it mean to reflect?

**Comment by jlperla** [comment]
>I'm confused by your explanation: my understanding is that one typically starts with a stochastic process, generates a continuous Fokker Plank PDE, then discretises the PDE with finite differences. It sounds like you want to jump directly from the stochastic process to the finite differences, and skip the continuous PDE?

This is not the Fokker Plank (which is the Kolgorov Forward equation) but rather it is a backwards equation (coming from things like the Feynman–Kac formula if it is a diffusion).  But you can write the Feynman-Kac formula more an arbitrary infinesimal generator.

Sorry if my notation and steps were unclear, but I was trying to skip a few steps in going from the stochastic process to the finite differences for brevity.  Think of this as starting with the infinitesimal generator of the process embedded in the hamilton jacobi-bellman equation. https://github.com/JuliaDiffEq/PDERoadmap/releases/download/v0.2.1/linear_operators_overview.pdf is more explicit on this for (non-jump) processes.

> In terms of the continuous PDE, the description of the jump process doesn't sound right, as it's described in terms of jumps on the (artificial) grid.

I was just trying to rig up an example where it was easy to figure out the stencil because the jump sizes lined up with the exact grid to keep things so I could write it out easily.  More generally you would have to give the continuous jump distributing and discretize that.  Sorry if I made things more confusing, but I think the stencil is something along those lines.

> I also don't understand "reflecting":

Reflecting in the sense of: https://math.nyu.edu/~varadhan/Spring11/topics16.pdf   But I still am having trouble finding how the finance guys write it down formally for jump-diffusions.

**Comment by jlperla** [comment]
> If the jump stays away from boundary, then shouldn't it still be Neumann conditions?

Probably. But the issue is that in general (and in most noncontrived examples) they wouldn't.

> It sounds like it would be convenient for people in finance to be able to specify things as an SDE and have the generation of the PDE happen automatically. Is this accurate?

I think finance people might like that (as a higher level library on top of the DiffEqOperators) but I don't always understand their tastes and applications. As an economist I would rather start with the infinitesimal generators of the stochastic process, which means composing differential operators. The other thing is that 90% of my applications are solving a stationary ODE in the spatial dimension rather than a PDE.
